dmscripts/export_framework_suppliers.py

```diff
@@ -0,0 +1,390 @@
+    mitigating_factors = [
+        ("MITIGATING 1", record['declaration'].get('mitigatingFactors', '')),
+        ("MITIGATING 2", record['declaration'].get('mitigatingFactors2', ''))
+    ]
+    return failed_questions + mitigating_factors
+
+
+def contact_details(record):
+    return [
+        ('contact_name', record['declaration'].get('primaryContact', '')),
+        ('contact_email', record['declaration'].get('primaryContactEmail', '')),
+    ]
+
+
+def service_counts(record):
+    return [
+        ('{}_{}'.format(status, lot), record['counts'][status][lot])
+        for status in DRAFT_STATUSES
+        for lot in LOTS
+    ]
+
+
+def write_csv(records, make_row, filename):
+    """Write a list of records out to CSV"""
+    def fieldnames(row):
+        return [field[0] for field in row]
+
+    writer = None
+
+    with open(filename, "w+") as f:
+        for record in records:
+            sys.stdout.write(".")
+            sys.stdout.flush()
+            row = make_row(record)
+            if writer is None:
+                writer = csv.DictWriter(f, fieldnames=fieldnames(row))
+                writer.writeheader()
+            writer.writerow(dict(row))
+
+
+class SuccessfulHandler(object):
+    NAME = 'successful'
+
+    def matches(self, record):
+        return record['onFramework'] is True
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class FailedHandler(object):
+    NAME = 'failed'
+
+    def matches(self, record):
+        return record['onFramework'] is False
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            failed_mandatory_questions(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class DiscretionaryHandler(object):
+    NAME = 'discretionary'
+
+    def matches(self, record):
+        return record['onFramework'] is None
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            discretionary_questions(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class MultiCSVWriter(object):
+    """Manage writing to multiple CSV files"""
+    def __init__(self, output_dir, handlers):
+        self.output_dir = output_dir
+        self.handlers = handlers
+        self._csv_writers = dict()
+        self._csv_files = dict()
+        self._counters = {
+            handler.NAME: 0 for handler in handlers
+        }
+
+    def write_row(self, record):
+        for handler in self.handlers:
+            if handler.matches(record):
+                self._counters[handler.NAME] += 1
+                row = handler.create_row(record)
+                return self.csv_writer(handler, row).writerow(dict(row))
+        raise Exception("record not handled by any handler")
```

Comment:
I would use a `ValueError` or maybe even a custom subclass - the problem with emitting bare `Exception`s is they are only catchable using `except Exception` which of course is troublesome.